src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -68,7 +69,23 @@ internal sealed partial class JsonClassInfo
             }
         }
 
+        private JsonConverter? _keyConverter;
+        public JsonConverter KeyConverter
+        {
+            get
+            {
+                if (_keyConverter == null)
```

Comment:
I wonder if a different approach would be to add new internal methods to JsonConverter to support keys. That way, we can just use the existing converters and don't need a new variable here, or add a new static ConcurrentDictionary.

Reply:
I see, you mean virtual methods that are implemented only for Dictionary converters, right? 
But then how do you determine which method call for each specific `TKey` depending on its type? 
You would have to use delegates (or something alike) for each type, i.e. one for `int, one for `Guid`, etc.

Reply:
I'm thinking we could add two new virtual methods to `JsonConverter<T>` like:
```c#
internal virtual string ReadAsString(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
{ throw new NotSupportedException(); }

internal virtual void WriteAsString(Utf8JsonWriter writer, TValue value, JsonSerializerOptions options)
{ throw new NotSupportedException(); }
```

which would be implemented by the Int32Converter, GuidConverter, etc.